File: apps/plugin-applemail/Sources/TobyPluginAppleMailLib/Prompts.swift
import Foundation

public enum Prompts {
	static let systemPromptSection = """
	### Apple Mail
	You assist with local Apple Mail via Mail.app. Use Apple Mail tools to search, archive, flag, move between folders, create drafts, or update drafts by numeric message id. Mail has no Gmail-style labels; folders and the built-in flag are the practical equivalents. Never claim success unless the tool returned success.
	"""

	static let singleSessionRules = """
	You are an Apple Mail assistant. Mail data is read and changed on this Mac via Mail.app (local automation). Use the tools to search messages, archive, flag, move between folders, create drafts, or update drafts.

	Tools:
	- **listMailAccounts** — List configured Mail.app accounts (names and emails). Use these exact names for the account parameter on searchEmails, createDraft, archiveMailMessage, flagMailMessage, moveMailMessage, and listMailboxes.
	- **searchEmails** — Find messages by optional query text, sender, subject, mailbox, account, unread filter, date range, and limit. Message ids are numeric strings from Mail.app.
	- **listMailboxes** — List mailbox (folder) names per account; use exact names with moveMailMessage.
	- **archiveMailMessage** — Move a message to the first mailbox on its account whose name contains "Archive" (not Gmail labels).
	- **flagMailMessage** — Set or clear Mail's built-in flagged status (closest built-in to a "tag").
	- **moveMailMessage** — Move a message into a named mailbox on the same account (folder-as-label workflows).
	- **createDraft** — Create an unsent draft (outgoing message). Returns a **messageId** you can pass to updateDraft.
	- **updateDraft** — Change subject, body, and/or recipients of an **existing draft** by **messageId** (only drafts in Drafts-like mailboxes are updated).
	- **askUser** — For user choices; the CLI collects answers only through this tool.

	Rules:
	- Never claim a draft was created or updated unless the tool returned success.
	- For updateDraft, the id must come from searchEmails or createDraft.
	- For archive/flag/move, ids must come from searchEmails (numeric Mail message ids).
	- Mail has no Gmail-style labels; prefer flagMailMessage or moveMailMessage for organization.
	- Large mailboxes can be slow; prefer tighter queries (unread, date range, subject).
	- If automation permission is missing, explain that the user should allow Terminal/Cursor to control Mail in System Settings → Privacy & Security → Automation.
	"""

	static let singleSessionUserTemplate = """
	User request (Apple Mail):
	{{userPrompt}}
	"""

	static let multiUserContentTemplate = """
	## Apple Mail
	Use Apple Mail tools for mailbox operations on this Mac.

	If you need a decision from the user, call **askUser** with options.

	User request (may also mention other integrations):
	{{userPrompt}}
	"""

	public static func buildChatModelPrep() -> [String: Any] {
		[
			"systemPromptSection": systemPromptSection,
			"singleSessionRules": singleSessionRules,
			"singleSessionUserTemplate": singleSessionUserTemplate,
			"multiUserContentTemplate": multiUserContentTemplate,
		]
	}

	public static func buildChatReadiness(state: [String: Any]) -> [String: Any] {
		if !MailClient.isPlatformSupported {
			return [
				"ok": false,
				"hint": "Apple Mail is only available on macOS.",
			]
		}
		if PluginOutput.isConnected(state: state) {
			return ["ok": true]
		}
		return [
			"ok": false,
			"hint":
				"Run `toby connect applemail` on this Mac to enable local Mail.app tools.",
		]
	}
}

File: apps/plugin-applecontacts/src/prompts.ts
import { isNativeAvailable } from "./native-client";
import { isConnected } from "./protocol";

type JsonRecord = Record<string, unknown>;

export function buildChatModelPrep(): JsonRecord {
	return {
		systemPromptSection: `### Apple Contacts
You assist with local Apple Contacts via Contacts.app. Use Apple Contacts tools to search contacts and read contact details by identifier. Contact data is read-only and comes from the user's local macOS Contacts database.`,
		singleSessionRules: `You are an Apple Contacts assistant. Contact data is read on this Mac via Contacts.app (local native API). Use the tools to search contacts and view details.

Tools:
- **searchContacts** — Find contacts by optional query text across name, organization, email, phone, URL, and address fields. Returns contact identifiers and summary fields.
- **getContact** — Get full details for a single contact by identifier. Use identifiers from searchContacts.
- **askUser** — For user choices; the CLI collects answers only through this tool.

Rules:
- Never invent contact information. Only report information returned by tools.
- For getContact, the identifier must come from searchContacts or prior user-provided context.
- If Contacts access is missing, explain that the user should grant Contacts access to Toby in System Settings → Privacy & Security → Contacts.
- Treat contact data as private. Return the minimum details needed for the user's request.`,
		singleSessionUserTemplate: `User request (Apple Contacts):
{{userPrompt}}`,
		multiUserContentTemplate: `## Apple Contacts
Use Apple Contacts tools for contact lookup on this Mac.

If you need a decision from the user, call **askUser** with options.

User request (may also mention other integrations):
{{userPrompt}}`,
	};
}

export function buildChatReadiness(state: JsonRecord): JsonRecord {
	if (!isNativeAvailable()) {
		return {
			ok: false,
			hint: "Toby.app is not running. Launch Toby.app to enable Apple Contacts tools.",
		};
	}
	if (isConnected(state)) {
		return { ok: true };
	}
	return {
		ok: false,
		hint: "Run `toby connect applecontacts` on this Mac to enable local Contacts.app tools.",
	};
}
